### PR TITLE
HB2-72 import routes from GPX files

### LINKED ISSUE #72: Import routes from GPX files
The new Swisstopo app only has a GPX feature. Routes created there are not available on the web. 

We should allow athletes to import GPX files.

### PR DESCRIPTION
**Pull request by sephii**
Fixes #72 .

It's still a WIP. Check the `TODO` and `FIXME` in the code to see what's left to be done.

### CODE REVIEW

File: homebytwo/routes/utils.py [hunk 1]
```diff
@@ -152,3 +152,17 @@ def get_places_within(point, max_distance=100):
...
+def get_distances(points):
+    """
+    Return a list with the distance of each point relative to the previous one in the list.
+    """
+
+    def get_relative_distances():
+        if points:
+            yield 0
+
+        yield from (p2.distance(p1) for p1, p2 in zip(points[1:], points))
```
Comment (drixselecta) [inline homebytwo/routes/utils.py:166]: nice one, using zip to shorten the list 🧠 🔥

File: homebytwo/routes/utils.py [hunk 2]
```diff
@@ -152,3 +152,17 @@ def get_places_within(point, max_distance=100):
...
+
+    return list(accumulate(get_relative_distances()))
```
Comment (drixselecta) [inline homebytwo/routes/utils.py:168]: I learned something today ❤️

File: homebytwo/importers/models/switzerlandmobilityroute.py [hunk 3]
```diff
@@ -132,7 +131,7 @@ def get_route_data(self, cookies=None, raw_route_json=None):
             )
 
             # create geom from lat, lng data columns
-            coords = [(lat, lng) for lat, lng in zip(data["lat"], data["lng"])]
+            coords = zip(data["lat"], data["lng"])
```
Comment (drixselecta) [inline homebytwo/importers/models/switzerlandmobilityroute.py:134]: Nice catch, I saw it last week and did not fix it. 😘 
There is even another useless list comprehension somewhere else in the code if I remember well.

Can you spot it?

**Review by drixselecta — COMMENTED** [review]
Thank you @sephii ❤️ !

File: homebytwo/importers/models/switzerlandmobilityroute.py (hunk 3 above)
Comment (drixselecta) [reply homebytwo/importers/models/switzerlandmobilityroute.py:134]: But it does seem to break the tests...

Comment (sephii) [reply homebytwo/importers/models/switzerlandmobilityroute.py:134]: Ah probably because it returns an iterator. Try surrounding it with `list()`.

Comment (drixselecta) [reply homebytwo/importers/models/switzerlandmobilityroute.py]: Works indeed. Spotted the other one as well: https://github.com/HomebyTwo/homebytwo/pull/73/commits/c72720afb05244f487d4e487ebba056405a92e50#diff-fecdbf464accb987a890345b98a2724dR118

File: homebytwo/importers/tests/test_gpx_import.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+try:
+    # Load LXML or fallback to cET or ET
+    import lxml.etree as mod_etree  # type: ignore
+except ImportError:
+    try:
+        import xml.etree.cElementTree as mod_etree  # type: ignore
+    except ImportError:
+        import xml.etree.ElementTree as mod_etree  # type: ignore
+
```
Comment (drixselecta) [inline homebytwo/importers/tests/test_gpx_import.py:20]: Remove this bit!

File: homebytwo/templates/routes/_route_card.html [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
 <li>
   <a href="{{ route.display_url }}">
     <div class="media">
       <div class="media__left">
-        <img src="{{ route.image.thumb.url|default:svg }}" alt="" height="45" width="45" />
+        {% if route.svg %}
+          <img src="{{ route.image.thumb.url|default:svg }}" alt="" height="45" width="45" />
+        {% endif %}
       </div>
       <div class="media__body relative-parent">
```
Comment (drixselecta) [inline homebytwo/templates/routes/_route_card.html:14]: Add svg for data_source homebytwo

**Comment by drixselecta** [comment]
Hi @sephii,

Thanks for your work. Here's what I tried to do to finish the job:
- [x] write tests for the GPX upload form and view
- [x] delete the `import_existing_route` view as it was the exact same as`routes:edit`
- [x] Display form errors on the page with the `importers/index.html` template
- [x] Prevent importing GPX with only one or no points. The application did not handle the resulting routes well.

There is still one `TODO` in the code, but for me, it would also work fine without:
- [ ] `templates/importers/gpx/_import_button.html`: hide the field and post using JS

Can you please look at it and tell me if you think that we are good to merge?

Forever yours,

Cédric

BTW: I don't think we should switch to 3D geometries.

**Pull request merged**